romtools/workflows/residual_evaluator.py

```diff
@@ -0,0 +1,184 @@
+
+    Returns:
+        `np.ndarray`: The full-order residual snapshots in tensor form.
+    """
+
+    run_directory_base = f"{absolute_run_directory}/res_"
+
+    all_residual_snapshots = []
+    n_vars = -1
+    n_x = -1
+    for index, full_model_dir in enumerate(full_state_directories):
+        # Read and project FOM snapshot
+        reduced_state = residual_evaluator.load_projected_full_solution(full_model_dir + "/" + state_filename)
+
+        # Set up corresponding directory
+        run_directory = f"{run_directory_base}{index}"
+        create_empty_dir(run_directory)
+
+        # Evaluate residual
+        residual_snapshot = residual_evaluator.evaluate_full_residual(run_directory, full_model_dir, reduced_state)
+
+        # check residual snapshot size and shape
+        assert residual_snapshot.ndim == 2
+        residual_snapshot = np.expand_dims(residual_snapshot, axis=2)
+        if n_vars == -1 and n_x == -1:
+            n_vars = residual_snapshot.shape[0]
+            n_x = residual_snapshot.shape[1]
+        assert residual_snapshot.shape[0] == n_vars
+        assert residual_snapshot.shape[1] == n_x
+
+        all_residual_snapshots.append(residual_snapshot)
+
+    # convert list to array
+    # does it kill memory usage to do this?
+    return np.concatenate(all_residual_snapshots, axis=2)
+
+
+def evaluate_and_load_unsteady_residual_snapshots(
+    residual_evaluator: UnsteadyResidualEvaluator, full_state_directories: Iterable[str], state_filename: str, absolute_run_directory: str
+) -> np.ndarray:
+    """
+    Core algorithm that takes a residual_evaluator, a list of unsteady full-order model
+    snapshot directories, and a snapshot_filename, and computes the corresponding
+    residual snapshots.
+
+    Args:
+        residual_evaluator (UnsteadyResidualEvaluator): steady residual evaluator we wish to use
+        full_state_directories (list[str]): list of directories containing full state data
+        state_filename (str): filename or base filename of file containing state data
+        absolute_run_directory (str): absolute path to base directory in which residuals are evaluated
+
+    Returns:
+        `np.ndarray`: The full-order residual snapshots in tensor form.
+    """
+
+    run_directory_base = f"{absolute_run_directory}/res_"
+
+    all_residual_snapshots = []
+    n_vars = -1
+    n_x = -1
+    for index, full_model_dir in enumerate(full_state_directories):
+        # Read and project FOM snapshots
+        reduced_states, times = residual_evaluator.load_projected_full_solutions(full_model_dir + "/" + state_filename)
+
+        # Set up corresponding directory
+        run_directory = f"{run_directory_base}{index}"
+        create_empty_dir(run_directory)
+
+        # Evaluate residuals
+        residual_snapshots = residual_evaluator.evaluate_full_residuals(run_directory, full_model_dir, reduced_states, times)
```

Comment:
Should we think about combining the steady and unsteady versions of this function?  It looks like the only difference is if the times argument and the dimension of the output.

Reply:
So just make the time stamps an optional argument? Right now we have two classes with two protocols, I don't see how going to one function helps unless we use inheritance somehow. Am I missing something?

Reply:
I'm thinking you could potentially just have one protocol that returns a 3d tensor (steady case would just have a length of 1 in that dimension).  I guess you'd still need a separate way to get the times if you did that...

I don't feel particularly strongly about this, but the two were just so similar that it seemed like there might be a way to combine them (or at least combine some aspects of them)

Reply:
i think these being interfaces it is fine to keep them separate but this does not prevent one to create a single concrete class that meets both